src/main/java/org/victorrobotics/frc/dtlib/drivetrain/trajectory/DTHolonomicTrajectoryGenerator.java

```diff
@@ -0,0 +1,226 @@
+
+import org.victorrobotics.frc.dtlib.drivetrain.DTAccelerationLimit;
+import org.victorrobotics.frc.dtlib.drivetrain.DTVelocityLimit;
+import org.victorrobotics.frc.dtlib.drivetrain.trajectory.DTHolonomicTrajectory.Point;
+import org.victorrobotics.frc.dtlib.math.geometry.DTVector2DR;
+import org.victorrobotics.frc.dtlib.math.spline.DTLinearSpline;
+import org.victorrobotics.frc.dtlib.math.spline.DTSpline;
+
+import java.util.ArrayList;
+import java.util.List;
+
+public class DTHolonomicTrajectoryGenerator {
+    private static final DTAccelerationLimit DEFAULT_ACCEL_LIMIT    = new DTAccelerationLimit(25, 20 * Math.PI, 0.02);
+    private static final DTVelocityLimit     DEFAULT_VELOCITY_LIMIT = new DTVelocityLimit(0.05, 5, 0.1, Math.PI * 4);
+
+    private static final double DEFAULT_DISTANCE_BETWEEN_POINTS = 0.02;
+    private static final double DEFAULT_MAX_DELTA_T             = 0.126;
+
+    private DTAccelerationLimit accelLimit;
+    private DTVelocityLimit     velocityLimit;
+    private double              maxDistance;
+    private double              maxDeltaU;
+
+    public DTHolonomicTrajectoryGenerator() {
+        accelLimit = DEFAULT_ACCEL_LIMIT;
+        velocityLimit = DEFAULT_VELOCITY_LIMIT;
+        maxDistance = DEFAULT_DISTANCE_BETWEEN_POINTS;
+        maxDeltaU = DEFAULT_MAX_DELTA_T;
+    }
+
+    public void setAccelLimit(DTAccelerationLimit limit) {
+        accelLimit = limit;
+    }
+
+    public void setVelocityLimit(DTVelocityLimit limit) {
+        velocityLimit = limit;
+    }
+
+    public void setMaxDistance(double distanceMeters) {
+        maxDistance = distanceMeters;
+    }
+
+    public DTHolonomicTrajectory generate(DTSpline<?> spline) {
+        Point[] points = generatePoints(spline);
+        computeDistances(points);
+        maximizeVelocities(points);
+        applyCentripetalConstraint(points);
+        applyAccelerationConstraint(points);
+        applyDecelerationConstraint(points);
+        computeTimes(points);
+        computeAccelerations(points);
+        return new DTHolonomicTrajectory(points);
+    }
+
+    private Point[] generatePoints(DTSpline<?> spline) {
+        List<Point> pointList = new ArrayList<>();
+        for (int i = 0; i < spline.length(); i++) {
+            Point start = createPoint(spline, i);
+            Point end = createPoint(spline, i + 1);
+            pointList.add(start);
+            recursiveSplit(pointList, spline, i, i + 1, start.position, end.position);
+        }
+        pointList.add(createPoint(spline, spline.length()));
+        return pointList.toArray(Point[]::new);
+    }
+
+    private void recursiveSplit(List<Point> pointList, DTSpline<?> spline, double minU, double maxU,
+            DTVector2DR startPos, DTVector2DR endPos) {
+        DTVector2DR displacement = endPos.clone();
+        displacement.subtract(startPos);
+        if (maxU - minU <= maxDeltaU && displacement.hypotenuse() <= maxDistance) {
```

Comment:
During execution, the plan is for points to be linearly interpolated based on the robot's current position and/or time (not sure which yet). That's the primary reason for `Point` implementing `Interpolateable`. This would be a slight improvement over the previous iteration, which just snapped the robot to the nearest point.

Reply:
The other thing that makes that difficult is that we have no idea how long the route will take until after the points are already locked in, and inserting more would require a full restart.